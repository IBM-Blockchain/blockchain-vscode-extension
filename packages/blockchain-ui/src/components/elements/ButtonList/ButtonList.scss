#root {


    #buttonList{
        .bx--btn{
            width: 140px;
            margin-right:16px;

            &:last-of-type{
                margin-right: 0;
            }
        }
    
        .bx--tab-content {
            width: 100%;
            padding-left: 0px;
            background-color: transparent;
        }

        
    }

}