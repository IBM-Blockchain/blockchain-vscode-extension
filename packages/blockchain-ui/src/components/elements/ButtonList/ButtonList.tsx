import React, { Component } from 'react';
import './ButtonList.scss';
import { Button } from 'carbon-components-react';

interface ButtonListProps {
    onProgressChange: (newIndex: number) => void;
    onDeployClicked: () => void;
    currentIndex: number;
    disableNext: boolean; // Disable Next button?
}

class ButtonList extends Component<ButtonListProps> {

    constructor(props: Readonly<ButtonListProps>) {
        super(props);

        this.incrementIndex = this.incrementIndex.bind(this);
        this.decrementIndex = this.decrementIndex.bind(this);
        this.deploy = this.deploy.bind(this);

    }

    render(): JSX.Element {
        const buttonList: JSX.Element[] = [];

        if (this.props.currentIndex === 0) {
            buttonList.push(<Button key='next' kind='primary' disabled={this.props.disableNext} onClick={this.incrementIndex}>Next</Button>);
        } else if (this.props.currentIndex === 1) {
            buttonList.push(<Button key='back' kind='secondary' onClick={this.decrementIndex}>Back</Button>);
            buttonList.push(<Button key='next' kind='primary' disabled={this.props.disableNext} onClick={this.incrementIndex}>Next</Button>);
        } else {
            buttonList.push(<Button key='back' kind='secondary' onClick={this.decrementIndex}>Back</Button>);
            buttonList.push(<Button key='next' kind='primary' onClick={this.deploy}>Deploy</Button>);
        }

        return (
            <div id='buttonList'>
                {buttonList}
            </div>
        );
    }

    incrementIndex(): void {
        const newIndex: number = this.props.currentIndex + 1;
        this.props.onProgressChange(newIndex);
    }

    decrementIndex(): void {
        const newIndex: number = this.props.currentIndex - 1;
        this.props.onProgressChange(newIndex);
    }

    deploy(): void {
        this.props.onDeployClicked();
    }
}

export default ButtonList;
