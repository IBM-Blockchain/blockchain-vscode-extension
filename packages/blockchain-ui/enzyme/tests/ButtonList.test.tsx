import React from 'react';
import renderer from 'react-test-renderer';
import { mount, ReactWrapper } from 'enzyme';
import chai from 'chai';
import sinon from 'sinon';
import sinonChai from 'sinon-chai';

import ButtonList from '../../src/components/elements/ButtonList/ButtonList';

chai.should();
chai.use(sinonChai);

// tslint:disable no-unused-expression

describe('ButtonList component', () => {
    let mySandBox: sinon.SinonSandbox;

    let onProgressChangeStub: sinon.SinonStub;
    let onDeployClickedStub: sinon.SinonStub;

    beforeEach(async () => {
        mySandBox = sinon.createSandbox();

        onProgressChangeStub = mySandBox.stub();
        onDeployClickedStub = mySandBox.stub();
    });

    afterEach(async () => {
        mySandBox.restore();
    });

    describe('render', () => {
        it('should render the expected snapshot', () => {
            const component: any = renderer
                .create(<ButtonList currentIndex={0} disableNext={false} onProgressChange={onProgressChangeStub} onDeployClicked={onDeployClickedStub} />)
                .toJSON();
            expect(component).toMatchSnapshot();
        });

        it('should render button for Step One', () => {
            const component: ReactWrapper<ButtonList> = mount(<ButtonList currentIndex={0} disableNext={false} onProgressChange={onProgressChangeStub} onDeployClicked={onDeployClickedStub} />);
            const buttonContainer: ReactWrapper<any> = component.at(0);

            const children: ReactWrapper<JSX.Element> = buttonContainer.children();
            const nextButton: ReactWrapper<JSX.Element> = children.childAt(0);
            nextButton.text().should.equal('Next');

            const props: any = nextButton.props();

            props.disabled.should.equal(false);
        });

        it('should disable button if no package selected', () => {
            const component: ReactWrapper<ButtonList> = mount(<ButtonList currentIndex={0} disableNext={true} onProgressChange={onProgressChangeStub} onDeployClicked={onDeployClickedStub} />);
            const buttonContainer: ReactWrapper<any> = component.at(0);

            const children: ReactWrapper<JSX.Element> = buttonContainer.children();
            const nextButton: ReactWrapper<JSX.Element> = children.childAt(0);
            const props: any = nextButton.props();
            props.disabled.should.equal(true);
        });

        it('should render buttons for Step Two', () => {
            const component: ReactWrapper<ButtonList> = mount(<ButtonList currentIndex={1} disableNext={false} onProgressChange={onProgressChangeStub} onDeployClicked={onDeployClickedStub} />);
            const buttonContainer: ReactWrapper<any> = component.at(0);

            const children: ReactWrapper<JSX.Element> = buttonContainer.children();
            const backButton: ReactWrapper<JSX.Element> = children.childAt(0);
            backButton.text().should.equal('Back');

            const nextButton: any = children.childAt(1);
            nextButton.text().should.equal('Next');
            const props: any = nextButton.props();
            props.disabled.should.equal(false);
        });

        it('should render buttons for Step Three', () => {
            const component: ReactWrapper<ButtonList> = mount(<ButtonList currentIndex={2} disableNext={false} onProgressChange={onProgressChangeStub} onDeployClicked={onDeployClickedStub} />);
            const buttonContainer: ReactWrapper<any> = component.at(0);

            const children: ReactWrapper<JSX.Element> = buttonContainer.children();
            const backButton: ReactWrapper<JSX.Element> = children.childAt(0);
            backButton.text().should.equal('Back');

            const deployButton: ReactWrapper<JSX.Element> = children.childAt(1);
            deployButton.text().should.equal('Deploy');
            const props: any = deployButton.props();
            props.disabled.should.equal(false);
        });
    });

    describe('incrementIndex', () => {
        it('should increment progress index', () => {
            const component: ReactWrapper<ButtonList> = mount(<ButtonList currentIndex={0} disableNext={false} onProgressChange={onProgressChangeStub} onDeployClicked={onDeployClickedStub} />);
            const instance: ButtonList = component.instance() as ButtonList;
            instance.incrementIndex();

            onProgressChangeStub.should.have.been.calledOnceWithExactly(1);
        });
    });

    describe('decrementIndex', () => {
        it('should decrement progress index', () => {
            const component: ReactWrapper<ButtonList> = mount(<ButtonList currentIndex={1} disableNext={false} onProgressChange={onProgressChangeStub} onDeployClicked={onDeployClickedStub} />);
            const instance: ButtonList = component.instance() as ButtonList;
            instance.decrementIndex();

            onProgressChangeStub.should.have.been.calledOnceWithExactly(0);
        });
    });

    describe('deploy', () => {
        it('should trigger deploy', () => {
            const component: ReactWrapper<ButtonList> = mount(<ButtonList currentIndex={1} disableNext={false} onProgressChange={onProgressChangeStub} onDeployClicked={onDeployClickedStub} />);
            const instance: ButtonList = component.instance() as ButtonList;
            instance.deploy();

            onDeployClickedStub.should.have.been.calledOnce;;
        });
    });

});
