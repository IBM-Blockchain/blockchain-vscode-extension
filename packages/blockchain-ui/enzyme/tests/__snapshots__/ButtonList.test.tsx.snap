// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonList component render should render the expected snapshot 1`] = `
<div
  id="buttonList"
>
  <button
    className="bx--btn bx--btn--primary"
    disabled={false}
    onClick={[Function]}
    tabIndex={0}
    type="button"
  >
    Next
  </button>
</div>
`;
